libs/common-types/src/lib.rs

```diff
@@ -0,0 +1,270 @@
+			PoolRole::TrancheInvestor(id, _) => self.tranche_investor.contains(id),
+		}
+	}
+
+	fn empty(&self) -> bool {
+		self.admin.is_empty() && self.tranche_investor.is_empty()
+	}
+
+	fn rm(&mut self, property: Self::Property) {
+		match property {
+			PoolRole::Borrower => self.admin.remove(AdminRoles::BORROWER),
+			PoolRole::LiquidityAdmin => self.admin.remove(AdminRoles::LIQUIDITY_ADMIN),
+			PoolRole::PoolAdmin => self.admin.remove(AdminRoles::POOL_ADMIN),
+			PoolRole::PricingAdmin => self.admin.remove(AdminRoles::PRICING_ADMIN),
+			PoolRole::MemberListAdmin => self.admin.remove(AdminRoles::MEMBER_LIST_ADMIN),
+			PoolRole::RiskAdmin => self.admin.remove(AdminRoles::RISK_ADMIN),
+			PoolRole::TrancheInvestor(id, delta) => self.tranche_investor.remove(id, delta),
+		}
+	}
+
+	fn add(&mut self, property: Self::Property) {
+		match property {
+			PoolRole::Borrower => self.admin.insert(AdminRoles::BORROWER),
+			PoolRole::LiquidityAdmin => self.admin.insert(AdminRoles::LIQUIDITY_ADMIN),
+			PoolRole::PoolAdmin => self.admin.insert(AdminRoles::POOL_ADMIN),
+			PoolRole::PricingAdmin => self.admin.insert(AdminRoles::PRICING_ADMIN),
+			PoolRole::MemberListAdmin => self.admin.insert(AdminRoles::MEMBER_LIST_ADMIN),
+			PoolRole::RiskAdmin => self.admin.insert(AdminRoles::RISK_ADMIN),
+			PoolRole::TrancheInvestor(id, delta) => self.tranche_investor.insert(id, delta),
+		}
+	}
+}
+
+impl<Now, MaxTranches, MinDelay, TrancheId, Moment>
+	TrancheInvestors<Now, MaxTranches, MinDelay, TrancheId, Moment>
+where
+	MaxTranches: Get<TrancheId>,
+	Now: Time<Moment = Moment>,
+	MinDelay: Get<Moment>,
+	Moment: PartialEq + PartialOrd + Saturating + Ord + Copy,
+	TrancheId: PartialEq + PartialOrd,
+{
+	pub fn empty() -> Self {
+		Self::default()
+	}
+
+	pub fn is_empty(&self) -> bool {
+		self.info.is_empty()
+	}
+
+	fn validity(&self, delta: Moment) -> Moment {
+		let now = Now::now();
+		let min_validity = now.saturating_add(MinDelay::get());
+		let req_validity = now.saturating_add(delta);
+		max(min_validity, req_validity)
+	}
+
+	pub fn contains(&self, tranche: TrancheId) -> bool {
+		if tranche >= self.max_tranches {
+			return false;
+		}
+
+		self.info
+			.iter()
+			.position(|info| info.tranche_id == tranche && info.permissioned_till >= Now::now())
+			.is_some()
+	}
+
+	pub fn remove(&mut self, tranche: TrancheId, delta: Moment) {
+		if tranche >= self.max_tranches {
+			return;
+		}
+
+		if let Some(index) = self.info.iter().position(|info| info.tranche_id == tranche) {
+			let valid_till = &self.info[index].permissioned_till;
+			let now = Now::now();
+
+			if *valid_till <= now {
+				// The account is already invalid. Hence no more grace period
+				return;
+			} else {
+				// Ensure that permissioned_till is at least now + min_delay.
+				self.info[index].permissioned_till = self.validity(delta);
+			}
+		};
+	}
+
+	pub fn insert(&mut self, tranche: TrancheId, delta: Moment) {
```

Comment:
since this is doing the upsert, I would lean towards removing the remove and rename this to upsert.
And error out like @offerijns suggested instead of setting the min_delay + now

Reply:
We can not really remove the remove as the remove also checks if the permission has already passed and errors out then. 

I changed it to the following (But can of course further adapt): 

**Insert**:
* delta + now < min_delay -> Error
* delta + now > min_delay -> update current permission or insert new object if not found. 
   I Think here we could discuss to only allow the increase of permissions_till and error out otherwise.
   
**Remove**:
* permissioned_till < now -> Error out
* delta + now < min_delay -> Error out
* Update permissioned_till otherwise

Reply:
Input wont be delta but rather a moment in the future. This will allow us to set what ever time we want without thinking about any time diffs.

It would look like this IMO

 ```
Insert:

provided_time < now + min_delay -> Error
provided_time >= now + min_delay -> update current permission or insert new object if not found.

Remove:
same as above
```

Reply:
@mustermeiszer any concerns on the above ?